### DISCUSSION TITLE
Is there a config to have `# KEYWORD(username):`?

### DISCUSSION THREAD

**Discussion by galenseilis** [folke/todo-comments.nvim#334] [Q&A]
I am excited about https://github.com/folke/todo-comments.nvim/issues/332! It should go well with Ruff's [TD002](https://docs.astral.sh/ruff/rules/missing-todo-author/).

In the mean time, is there currently a way to support this in config?

I made an attempt, but in retrospect I see that trying to update the REGEX is insufficient. I expect I am missing things specifying the allowed patterns and what colour they should get...

```lua
-- init.lua
{
    'folke/todo-comments.nvim',
    event = 'VimEnter',
    dependencies = { 'nvim-lua/plenary.nvim' },
    opts = {
      signs = false,
      search = {
        command = 'rg',
        args = {
          '--color=never',
          '--no-heading',
          '--with-filename',
          '--line-number',
          '--column',
        },
        pattern = [[\b(KEYWORDS)(\(\w+\))?:]], -- updated pattern to match KEYWORDS and KEYWORDS(name)
      },
    },
  }
```

**Comment by inkfin** [comment]
Hi, I came across this problem and unfortunately it seems that Vim regex has some issues that prevent the ? quantifier from working properly.

I've tried @galenseilis answer, but it failed to match TODO with parentheses. I also tried a few other approaches but wasn’t lucky either, so I ended up doing this instead:

```lua
highlight = {
    pattern = {
        [[.*<(KEYWORDS)\s*:]],              -- TODO: this matches
        [[.*<(KEYWORDS)\s*\(\w*\)\s*:]],    -- TODO(author): this also matches
    },
},
search = { pattern = [[\b(TODO)\s*(\(\w*\))?\s*:]] },
```

Fortunately, the rg regex works fine — hope this helps anyone running into the same issue!

**Comment by mattness** [comment]
I also ran into this problem, and tried the patterns from both @galenseilis and @inkfin.  Inkfin's pattern list got me close, but didn't highlight the entire `(user)` part of the comment:
<img width="211" height="150" alt="image" src="https://github.com/user-attachments/assets/6cce475b-d82f-46bf-8cef-2dd55857cb09" />

I ended up settling on this pattern that both correctly highlights and correctly populates the quickfix list:
```
highlight = {
  pattern = [[.*<((KEYWORDS)(\(\w+\))?)\s*:]],
},
search = {
  pattern = [[\b(KEYWORDS)(?:\(\w+\))?\s*:]],
},
```
<img width="204" height="149" alt="image" src="https://github.com/user-attachments/assets/9ccf55ff-00e1-4132-adf8-a9b76ea53b19" /> 
<img width="352" height="146" alt="image" src="https://github.com/user-attachments/assets/7d33466a-ad90-4d7c-a18e-601e269e95cd" />

The key difference here is wrapping an additional capture group around both the `(KEYWORDS)` group and the `(\(\w+\))?` group for the `(user)` part. This is necessary because the highlighter expects the keyword (TODO, NOTE, etc) to be in the second capture _or_ the third capture if a third is present and without the wrapper the third capture group ends up containing `(user)` which isn't a keyword. The wrapping capture group forces the `(KEYWORDS)` group to always be the third group, even when`(user)` is absent, making the pattern work for both cases.

See here:
https://github.com/folke/todo-comments.nvim/blob/19d461ddd543e938eb22505fb03fa878800270b6/lua/todo-comments/highlight.lua#L50-L56

So, for the capturing part of the pattern - `((KEYWORDS)(\(\w+\))?)` - the `m` table contains the following:
- `m[2]` = The contents of the outer capture group: `(KEYWORDS)(\(\w+\))?`
- `m[3]` = The contents of the first inner capture group: `KEYWORDS`
- `m[4]` (unused) = The contents of the second inner capture group: `\(\w+\)`

The following table illustrates the outcome:

Input | m[2] / match | m[3] | m[4] | Resulting kw
-- | -- | -- | -- | --
`-- TODO: msg` | TODO | TODO | "" | ✅ TODO
`-- TODO(user): msg` | TODO(user) | TODO | (user) | ✅ TODO

**Reply by mazunki** [reply]
Thanks! I changed `\w` with my username to only highlight my own TODOs:

```lua
opts = {
  highlight = {
    pattern = [[.*<((KEYWORDS)(\(mazunki\))?)\s*:]],
  },
  search = {
    pattern = [[\b(KEYWORDS)(?:\(\w+\))?\s*:]],
  },
}
```

**Comment by tgaldi** [comment]
If you don't want to highlight inside the ():

            pattern =
            {
                [[.*<(KEYWORDS)\s*:]],
                [=[.*<(KEYWORDS)\s*\(]=],
            }
